File: BCKEND/TypeScript/BancoFinanceiro/src/Model/Conta.ts
import { Agencia } from "./Agencia";
import { Cliente } from "./Cliente";
import { TipoTransacao, Transacao } from "./Transacao";

export abstract class Conta {

    private cliente: Cliente;
    private numero: number;
    private agencia: Agencia;
    private dataCriacao: Date;
    private saldo: number;
    private status: boolean;
    private transacoes: Transacao[];

    constructor(cliente: Cliente, numero: number, agencia: Agencia) {
        this.cliente = cliente;
        this.numero = numero;
        this.agencia = agencia;
        this.dataCriacao = new Date();
        this.saldo = 0;
        this.status = true;
        this.transacoes = [];
    }

    public getCliente(): Cliente {
        return this.cliente;
    }

    public getNumero(): number {
        return this.numero;
    }

    public getAgencia(): Agencia {
        return this.agencia;
    }

    public getDataCriacao(): Date {
        return this.dataCriacao;
    }

    public getSaldo(): number {
        return this.saldo;
    }

    public isAtiva(): boolean {
        return this.status;
    }

    public getTransacoes(): Transacao[] {
        return this.transacoes;
    }

    //depositar
    public depositar(valor: number): boolean {
        if (valor <= 0 || !this.status) {
            console.log("Valor inválido ou conta inativa.");
            return false;
        }else {
            this.saldo += valor;
            const transacao = new Transacao(TipoTransacao.DEPOSITO, valor, 
                "+");
            this.transacoes.push(transacao);
            return true;
        }
    }
    //sacar
    public sacar(valor: number): boolean {
        if (valor <= 0 || valor > this.saldo || !this.status) {
            console.log("Valor inválido, saldo insuficiente ou conta inativa.");
            return false;
        }else {
            this.saldo -= valor;
            const transacao = new Transacao(TipoTransacao.SAQUE, valor, 
                "-");
            this.transacoes.push(transacao);
            return true;
        }
    }

    //transferir
    public transferir(valor: number, contaDestino: Conta): boolean {
        if (valor <= 0 || valor > this.saldo || !this.status) {
            console.log("Valor inválido, saldo insuficiente ou conta inativa.");
            return false;
        }else {     
            this.saldo -= valor;
            contaDestino.saldo += valor;
            const transacaoSaida = new Transacao(TipoTransacao.TRANSFERENCIA, valor, 
                "-", contaDestino.cliente);
            this.transacoes.push(transacaoSaida);
            const transacaoEntrada = new Transacao(TipoTransacao.TRANSFERENCIA, valor, 
                "+", this.cliente);
            contaDestino.transacoes.push(transacaoEntrada);
            return true;
        }
    }

    //pagar
    public pagar(valor: number, descricao: string): boolean {
        if (valor <= 0 || valor > this.saldo || !this.status) {
            console.log("Valor inválido, saldo insuficiente ou conta inativa.");
            return false;
        }else {     
            this.saldo -= valor;
            const transacao = new Transacao(TipoTransacao.PAGAMENTO, valor, 
                "-", undefined, descricao);
            this.transacoes.push(transacao);
            return true;
        }
    }

    //imprimir extrato
    public gerarExtrato(): void {
        console.log(`\nExtrato da Conta ${this.numero} - Agência ${this.agencia.getNumero()}`);
        console.log(`Cliente: ${this.cliente.getNome()}`);
        console.log("Transações:");
        if (this.transacoes.length === 0) {
            console.log("Nenhuma transação realizada.");
        } else {
            this.transacoes.forEach((transacao, index) => {
                console.log(`${transacao.toString()}`);
            });
        }
        console.log(`Saldo Atual: R$ ${this.saldo.toFixed(2)} \n`);
    }

}

File: BCKEND/JAVA/BancoFinanceiro/src/Model/Conta.java
package Model;

import java.util.ArrayList;
import java.util.Date;

public class Conta {
	
	private Cliente cliente;
	private int numero;
	private Date dataCriacao;
	private Agencia ag;
	private double saldo;
	private boolean status;
	private ArrayList<Transacao> transacoes;
	
	public Conta(Cliente cliente, int numero, Agencia ag) {
		this.cliente = cliente;
		this.numero = numero;
		this.ag = ag;
		this.dataCriacao = new Date();
		this.saldo = 0.0;
		this.status = true;
		this.transacoes = new ArrayList<Transacao>();
		
	}

	public Conta(Cliente cliente, Date dataCriacao, Agencia ag, double saldo) {
		super();
		this.cliente = cliente;
		this.dataCriacao = dataCriacao;
		this.ag = ag;
		this.saldo = saldo;
		this.status = true;
		this.transacoes = new ArrayList<Transacao>();
	}
	
	//sacar
	public boolean sacar(double valor) {
		if (valor >= 0) {
			if (this.saldo >= valor) {
				this.saldo -= valor; // this.saldo = this.saldo - valor
				Transacao transacao = new Transacao(TipoTransacao.Saque, 
						valor, '-');
				this.transacoes.add(transacao);
				return true;
			} else {
				System.out.println("Saldo insuficiente!");
				return false;
			}			
		} else {
			System.out.println("Valor inválido!");
			return false;
		}		
	} 
	
	//depositar
	public boolean depositar(double valor) {
		if (valor >= 0) {
			this.saldo += valor;
			Transacao transacao = new Transacao(TipoTransacao.Depósito, 
					valor, '+');
			this.transacoes.add(transacao);
			return true;
		} else {
			System.out.println("Valor inválido!");
			return false;
		}
	}
	
	//transferir
	public boolean transferir(double valor, Conta contaDestino) {
		if (valor > 0 && contaDestino != null && contaDestino.status) {
			if(this.saldo >= valor) {
				this.saldo -= valor;
				Transacao transSaida = new Transacao(TipoTransacao.Transferência,
						valor, contaDestino.cliente, '-');
				this.transacoes.add(transSaida);
				contaDestino.saldo += valor;
				Transacao transDestino = new Transacao(TipoTransacao.Transferência, 
						valor, this.cliente, '+');
				contaDestino.transacoes.add(transDestino);
				return true;
			} else {
				System.out.println("Saldo insuficiente!");
				return false;
			}
		} else {
			System.out.println("Valores inválidos!");
			return false;
		}
	}
	
	//pagar
	public boolean pagar(double valor, String infoPagmento) {
		if (valor > 0 && infoPagmento != null ) {
			if (this.saldo >= valor) {
				this.saldo -= valor;
				Transacao transacao = new Transacao(TipoTransacao.Pagamento, 
						valor, infoPagmento, '-');
				this.transacoes.add(transacao);
				return true;
			} else {
				System.out.println("Saldo insuficiente!");
				return false;
			}
		} else {
			System.out.println("Valores inválidos!");
			return false;
		}
	}
	
	//imprimir extrato
	public String gerarExtrato() {
		String extrato = ".:: Extrato da Conta " + this.numero + " ::.\n";
		extrato += "Cliente: " + this.cliente.getNome() + ", CPF: " + 
				this.cliente.getCpf() + "\n" + "Agência " + this.ag.getNome()
				+ ", Nº " + this.getNumero() + "\n";
		for (Transacao transacao : transacoes) {
			extrato += transacao + "\n";
		}
		extrato += "Salado atual: R$ " + this.saldo;		
		return extrato;
	}

	public Agencia getAg() {
		return ag;
	}

	public void setAg(Agencia ag) {
		this.ag = ag;
	}

	public boolean isStatus() {
		return status;
	}

	public void setStatus(boolean status) {
		this.status = status;
	}

	public Cliente getCliente() {
		return cliente;
	}

	public Date getDataCriacao() {
		return dataCriacao;
	}

	public double getSaldo() {
		return saldo;
	}
	
	public int getNumero() {
		return numero;
	}

	public ArrayList<Transacao> getTransacoes() {
		return transacoes;
	}	
	
}
